### PR TITLE
feat: partition filters for TargetLevel version and a complete test

### PR DESCRIPTION
**Pull request by NGA-TRAN**
The Another PR  for https://github.com/influxdata/idpe/issues/16983 and a part of https://github.com/influxdata/influxdb_iox/pull/6717

This is the last piece of implementation Compaction Algorithm #2: `TargetLevel`

More tests needed but we can now turn it on test clusters

### CODE REVIEW

File: compactor2/src/components/hardcoded.rs [hunk 1]
```diff
@@ -332,3 +331,17 @@ fn version_specific_upgrade_split(config: &Config) -> Arc<dyn FilesSplit> {
         )),
     }
 }
+
+// filter files of a partition we do not compact
+fn version_specific_files_filter(config: &Config) -> Arc<dyn FilesFilter> {
+    match config.compact_version {
```
Comment (NGA-TRAN) [inline compactor2/src/components/hardcoded.rs:337]:
```suggestion
    match config.compact_version {
    // AllAtOnce filters our L2 files
```

File: compactor2/src/components/hardcoded.rs [hunk 2]
```diff
@@ -332,3 +331,17 @@ fn version_specific_upgrade_split(config: &Config) -> Arc<dyn FilesSplit> {
...
+        AlgoVersion::AllAtOnce => Arc::new(FilesFilterChain::new(vec![Arc::new(
+            PerFileFilesFilter::new(AndFileFilter::new(vec![Arc::new(
+                LevelRangeFileFilter::new(
+                    CompactionLevel::Initial..=CompactionLevel::FileNonOverlapped,
+                ),
+            )])),
+        )])),
```
Comment (NGA-TRAN) [inline compactor2/src/components/hardcoded.rs:344]:
```suggestion
        )])),
        // TargetLevel does not filter any files of the partition
```

File: compactor2/src/config.rs [hunk 3]
```diff
@@ -104,6 +104,10 @@ pub struct Config {
 
     /// Version of the compaction algorithm.
     pub compact_version: AlgoVersion,
+
+    /// Minimum number of L1 files to compact to L2
+    /// This is to prevent too many small files
+    pub min_num_l1_files_to_compact: usize,
```
Comment (alamb) [inline compactor2/src/config.rs:110]: I suggest in the future a better setting might be "target l2 size" which is perhaps closer to the intent (get a few large l2 files)

File: compactor2/src/test_util.rs [hunk 4]
```diff
@@ -525,6 +527,20 @@ impl TestSetup {
         assert_eq!(file.table_id, self.table.table.id);
         self.table.read_parquet_file(file).await
     }
+
+    /// Set compact version
+    pub fn set_compact_version(&mut self, compact_version: AlgoVersion) {
+        let mut config = Arc::get_mut(&mut self.config).unwrap();
+        config.compact_version = compact_version;
+        self.config = Arc::new(config.clone());
```
Comment (alamb) [inline compactor2/src/test_util.rs:535]: I don't think you need to both modify the arc and then also clone it 

```suggestion
        let mut config = Arc::get_mut(&mut self.config).unwrap();
        config.compact_version = compact_version;
```

File: compactor2/src/test_util.rs [hunk 5]
```diff
@@ -525,6 +527,20 @@ impl TestSetup {
...
+    }
+
+    /// set min_num_l1_files_to_compact
+    pub fn set_min_num_l1_files_to_compact(&mut self, min_num_l1_files_to_compact: usize) {
+        let mut config = Arc::get_mut(&mut self.config).unwrap();
+        config.min_num_l1_files_to_compact = min_num_l1_files_to_compact;
+        self.config = Arc::new(config.clone());
```
Comment (alamb) [inline compactor2/src/test_util.rs:542]:
```suggestion
```

**Review by alamb — APPROVED** [review]
This code looks good to me.

I still don't quite understand how the following works:
1. A partition has a very large number of L0 files (that can't all be compacted in a single plan)
2. A partition has overlapping large files

Maybe this is worth adding test coverage for?

File: compactor2/src/compactor_tests.rs [hunk 6]
```diff
@@ -157,6 +159,129 @@ mod tests {
...
+                (1, time_1_minute_future),
+                (2, time_2_minutes_future),
+                (3, time_5_minutes_future),
+                (4, time_3_minutes_future),
+                (5, time_5_minutes_future),
+                (6, time_2_minutes_future),
+            ]
+        );
+
+        // compact
```
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply compactor2/src/compactor_tests.rs:215]: Sounds great. We definitely need to add more tests and can't wait to use your new display

**Comment by NGA-TRAN** [comment]
@alamb 

> A partition has a very large number of L0 files (that can't all be compacted in a single plan)

We do not solve this in this Version2. What we do in this version is to not compact overlapping files which is more important to keep the compaction efficient and as a result won't leave so many L0s files around. 
I am about to open many tickets, one of which is to split many L0s files into many groups and compact them in parallel into L0s. This case, we will be able to compact partitions that are behind. This will be in Version3 which we can start as early as next week but we need Version2 first

> A partition has overlapping large files

We do not solve this in Version2 either but what we do in Version2 is to prevent this situation from happening. If it happens, we have to work on different solution to solve it as above

**Comment by alamb** [comment]
> not compact non-overlapping files which is more important to keep the compaction efficient and as a result won't leave so many L0s files around.

I would like to discuss the tradeoffs involved -- I understand the compaction efficiency is important for high ingest clusters, but avoiding a large number of files per partition is quite critical for users too. 

> I am about to open many tickets, one of which is to split many L0s files into many groups and compact them in parallel into L0s. This case, we will be able to compact partitions that are behind. This will be in Version3 which we can start as early as next week but we need Version2 first

Can we please start with a document about what the goals of the Version3 algorithm are. For example, I think it is critcally important to avoid large numbers of L0 files as well as ensure the hot leading edge of inserts to be compacted

**Comment by NGA-TRAN** [comment]
Yes, Version3 is critical but Version2 is more critical and a must-foundation for Version3
Yes, I am put the goals in Version3 and we should start brainstorm and work on it asap since now we have Version2 available

**Pull request merged**